pallets/confidential/src/lib.rs

```diff
@@ -0,0 +1,164 @@
+// This file is part of the Polymesh distribution (https://github.com/PolymathNetwork/Polymesh).
+// Copyright (c) 2020 Polymath
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, version 3.
+
+// This program is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+#![cfg_attr(not(feature = "std"), no_std)]
+#![feature(box_syntax)]
+
+use polymesh_common_utilities::{identity::Trait as IdentityTrait, Context};
+use polymesh_primitives::{AccountId, IdentityId, Ticker};
+use polymesh_primitives_derive::{SliceU8StrongTyped, VecU8StrongTyped};
+
+use pallet_identity as identity;
+
+use bulletproofs::RangeProof;
+use cryptography::asset_proofs::range_proof::{prove_within_range, verify_within_range};
+use curve25519_dalek::{ristretto::CompressedRistretto, scalar::Scalar};
+
+use codec::{Decode, Encode};
+use core::convert::TryFrom;
+use frame_support::{
+    debug, decl_error, decl_event, decl_module, decl_storage, dispatch::DispatchResult,
+    weights::SimpleDispatchInfo,
+};
+use frame_system::{self as system, ensure_signed};
+use sp_std::prelude::*;
+
+pub mod rng;
+pub use rng::native_rng;
+
+#[derive(Encode, Decode, Clone, Debug, Default, PartialEq, Eq, SliceU8StrongTyped)]
+pub struct RangeProofInitialMessageWrapper(pub [u8; 32]);
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, VecU8StrongTyped)]
+pub struct RangeProofFinalResponseWrapper(pub Vec<u8>);
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub struct TickerRangeProof {
```

Comment:
Range proofs are for account balances, where the value of a u32 can change, and you'd want to prove that it's within a range. To prove that a ticker is within a set of valid tickers, you should use Membership proofs.

I guess you put this file together as a way to start the integration. Keep in mind that PolyMesh should make proofs via the `mercat` library, and shouldn't need to make its own proofs. If there is a logic missing, it should be implemented in the `mercat` library.

Reply:
Yes, as you said, that is just to double-check any technical issue with the integration (a.k.a. wasm issues :smile: )
The content of this module is used just for testing that integration.